### ISSUE TITLE
1364 Field 'description' doesn't have a default value error while creating new category

### ISSUE THREAD

**Issue by szepty-ziemi** [JoomGalleryfriends/JG4-dev#4]
GIVEN
Browse to Backend -> Components -> JomGallery -> Category Manager
WHEN
Clicking _New_ button to create new category
THEN
Error is displayed: An error has occurred 1364 Field 'description' doesn't have a default value
BUT
Category creation form should be displayed

![Screenshot_1](https://user-images.githubusercontent.com/60470150/159166831-a31104e2-cc0d-4b72-84d5-0c81d3375010.png)

This issue apply to following columns in _joomgallery_categories table

description
params
created_time
modified_time
metadesc
metakey

It seems that those columns have default value set to _None_ while they should rather have _Null_. So setting:

`ALTER TABLE `xxx_joomgallery_categories` CHANGE `metakey` `metakey` TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NULL DEFAULT NULL; `

Fixed the issue on my installation.

Joomla version and settings:

![obraz](https://user-images.githubusercontent.com/60470150/159167026-45c0b662-4b81-46e9-b01b-f6b2c7cc8be2.png)

JoomGallery version: https://github.com/JoomGalleryfriends/JG4-dev/pull/1

**Comment by Elfangor93** [comment]
@szepty-ziemi Please add the steps to reproduce the bug as described here:
https://github.com/JoomGalleryfriends/JG4-dev#testing

... I can not reproduce this bug with my development environment ...

**Comment by szepty-ziemi** [comment]
Steps to replicate (they are already described in Gherkin):

1. Need to have Joomla 4 installation, without JoomGallery installed.
2. Download JoomGallery as ZIP from main branch
3. Install it (there could be an issue with installing)
4. Browse to Category Manager
5. Click button New

I also recorded a video for you, that shows above steps: https://www.youtube.com/watch?v=mgdAtCYdk9U

**Comment by Elfangor93** [comment]
I am most interested in:

**System information**
- PHP-Version
- Database type and version

**Comment by szepty-ziemi** [comment]
![obraz](https://user-images.githubusercontent.com/60470150/160255941-ca33ef86-ce4b-4cdb-b3dd-f05027922bdb.png)

More details: https://pastebin.com/SmkwRSst

**Comment by Elfangor93** [comment]
I still can not reproduce the error on my system. Please post your current SQL mode:
https://stackoverflow.com/questions/15438840/mysql-error-1364-field-doesnt-have-a-default-values

You can get your current SQL mode by perfoming `select @@GLOBAL.sql_mode` with your database management tool, such as phpMyAdmin.

**Comment by Elfangor93** [comment]
> It seems that those columns have default value set to None while they should rather have Null.

No. Columns with data type "text" should not have any default value.

![grafik](https://user-images.githubusercontent.com/39154009/160343299-4b8dac8d-4371-4bfc-a2d2-41f320859ed0.png)
source: https://dev.mysql.com/doc/refman/5.7/en/data-type-defaults.html

**Comment by szepty-ziemi** [comment]
@Elfangor93 Result of the query:

![image](https://user-images.githubusercontent.com/60470150/160345184-5581d053-5b82-4ee4-8243-40fd5ac8151d.png)
 I don't have STRICT_TRANS_TABLES value.

There is other way around to fix this issue, but I'm not sure if this would be valid. I edited file ROOT/administrator/components/com_joomgallery/src/Table/CategoryTable.php in line 378 and added:

`->set('description = ' . $db->quote(''))`

This allowed me to skip the error and move to next one for params:

![image](https://user-images.githubusercontent.com/60470150/160347347-e9d79160-fafe-43c4-a0b0-7a85302ff5d1.png)

![image](https://user-images.githubusercontent.com/60470150/160347618-a82519c8-7318-4882-a53d-4806d09b4a21.png)

**Comment by Elfangor93** [comment]
Interesting. The method CategoryTable::addRoot() should not be executed, since the category root gets created within the installation script.
https://github.com/JoomGalleryfriends/JG4-dev/blob/046690bc8034cceba042dbb5286149adf8dbf9d2/script.php#L290

Did you get any error during installation?

**Comment by Elfangor93** [comment]
Is your Joomla a clean Joomla 4.1 installation or was it migrated from Joomla 3.10?

**Comment by szepty-ziemi** [comment]
This is very cleam installation of Joomla 4, with the the only purpose to test JoomGallery. So installation could be the root cause of the issue... Joomla installatipn went smooth, without any errors. What I could do more to debug this?

**Comment by MrMusic** [comment]
I did a new install of Joomla 4.1.0 and installed the current JG4-dev main branch.
Then created categories without error.
Joomla 4.1.0
PHP 8.1.2
MariaDB 10.4.22
SQL strict mode: with or without 'strict_all_tables' makes no difference

**Comment by szepty-ziemi** [comment]
I'm providing SU credentials to my instance, if it will be any help:

u: joomgallery
p: ABjGxHyf8Q1CVg

**Comment by MrMusic** [comment]
@szepty-ziemi  Your test installation does not contain any entry for the JoomGallery configuration.
Try to uninstall the JoomGallery and then install the current JG4-dev branch again.

**Comment by szepty-ziemi** [comment]
@MrMusic I've tried few times... all in vain.

So I've checked this file: https://github.com/JoomGalleryfriends/JG4-dev/blob/main/script.php and it raised some question marks.

If I comment this (also similar lines for config etc):

```
    if (!$table->bind($data))
    {
      Factory::getApplication()->enqueueMessage(Text::_('Error bind category'), 'error');

      return false;
    }
    if (!$table->store($data))
    {
      Factory::getApplication()->enqueueMessage(Text::_('Error store category'), 'error');

      return false;
    }
```
Installation passes but not default category or configuration  is installed.

Then we have:

```
// Set level and parent_id
    $fields = array(
      $db->quoteName('parent_id') . ' = ' . $db->quote($data['parent_id']),
      $db->quoteName('level') . ' = ' . $db->quote($data['level'])
    );
    $conditions = array (
      $db->quoteName('alias') . ' = ' . $db->quote($data['alias'])
    );

    // insert to database
    $query = $db->getQuery(true);
    $query->update($db->quoteName('#__joomgallery_categories'))->set($fields)->where($conditions);
    $db->setQuery($query);
    $db->execute();
```

We use update here... but I think that we are creating parent category for the first time. Shouldn't it be insert there?
Also this piece of code is not present for _Add a category to the ´#__joomgallery_configs´ table_ is it okay?

I admit that I'm not an expert and I'm just trying to understand this issue... so trying some things.

I'm downloading a zip file from GitHhub as described and installing usual way... no juice. You also have access so you may try to do the same. 

I'm afraid that without solving this issue on my instance I won't be able to test commits correctly.

**Comment by Elfangor93** [comment]
I can confirm the error during installation on this system:
![grafik](https://user-images.githubusercontent.com/39154009/161307397-7a3dcbe9-223d-433f-aaff-ae42b08721f6.png)

On my development environment built in windows i dont get the error. I think the error is related to the linux as operation system.

**Comment by Elfangor93** [comment]
I guess the error is based on the way I registered the Table classes. I guess the paths are created wrong on linux systems...
https://github.com/JoomGalleryfriends/JG4-dev/blob/046690bc8034cceba042dbb5286149adf8dbf9d2/script.php#L294

**Comment by Elfangor93** [comment]
@szepty-ziemi please try if you still get the error with PR #10

**Comment by szepty-ziemi** [comment]
It works as a charm :)

![obraz](https://user-images.githubusercontent.com/60470150/161390042-9d2c6fb3-00fa-456e-8448-c1fd6b764eb6.png)

![obraz](https://user-images.githubusercontent.com/60470150/161390060-d7410457-0d1e-4f4d-9115-59fa3147c6c5.png)

![obraz](https://user-images.githubusercontent.com/60470150/161390075-149926ff-8d70-40c8-b23f-0d01279ab37d.png)

**Comment by Elfangor93** [comment]
Yeah, it was quite a pain in the ass to solve. I hope my hack does not have any side effects ;-)

**Comment by szepty-ziemi** [comment]
@Elfangor93 okay, so let's thing about regression test cases. What should be checked? What could be influenced by this change?
Installation: check on Windows servers and Linux servers. PHP version?

**Comment by Elfangor93** [comment]
### Tests to check install/update/uninstall
**Checks after Installation:**
- No errors during installation
- Default category created (Uncategorized)
- Default image types created (original, detail, thumbnail)
- Folder structure correctly created (/images/joomgallery/<imagetype>/uncategorized)
- Global configuation set created
- Ability to create and delete categories, images, configuration sets and image types

**Checks after Update** (installing the component again):
- No errors during update
- Ability to create and delete categories, images, configuration sets and image types

**Checks after Uninstallation:**
- No errors during uninstallation
- DB: joomgallery related tables deleted
- Table #__assets: all joomgallery related assets deleted
- Files and folders deleted (/images/joomgallery/)

### Testing everything with:
- PHP 7.2.x / 7.3.x / 7.4.x / 8.0.x and 8.1.x
- MySQL 5.x / MySQL 8.x / MariaDB 10.x
- Apache / NginX
- Linux / Windows

### Mindest system voraussetzungen
- Joomla 4.0
- PHP 7.3.0
If the component is installed with a lower Joomla or PHP version, the installation should stop and and an error message will be shown.

**Comment by Elfangor93** [comment]
>MySQL 5.x / MySQL 8.x / MariaDB 10.x

@MrMusic Which database types do you think should be supported by Joomla 4.x?

**Comment by Elfangor93** [comment]
Issue fixed with PR #10.
